### PR TITLE
promise combinators

### PR DESCRIPTION
**Pull request by avillega**

### CODE REVIEW

**Comment by avillega** [comment]
Putting this draft PR out there to get some reviews started. Still need to add tests and rebase main on this one.

**Comment by avillega** [comment]
This change is ready for review.

**Review by dfarr — DISMISSED** [review]
This looks fantastic! Combinators are just really cool

File: src/resonate/actions.py [hunk 1]
```diff
@@ -59,3 +68,180 @@ def with_options(
...
+            retry_policy=retry_policy if retry_policy is not None else never(),
+        )
+        return self
+
+    @abstractmethod
+    def done(self) -> bool:
+        pass
+
+    @abstractmethod
+    def result(self) -> Result[T, Exception]:
```
Comment (Tomperez98) [inline src/resonate/actions.py:109]: All methods in the `actions` classes are meant to be used by the user. I don't really love the idea of having `result()` and `done()` here. It seems confusing to me, I don't know what you think

File: src/resonate/actions.py [hunk 2]
```diff
@@ -59,3 +68,180 @@ def with_options(
...
+
+        Returns:
+            bool: True if all promises are complete, False otherwise.
+        """
+        if len(self.promises) == 0:
+            return True
+
+        return all(p.done() for p in self.promises)
+
+    def result(self) -> Result[list[Any], Exception]:
```
Comment (Tomperez98) [inline src/resonate/actions.py:137]: The typing here seems wrong. Isn't this supposed to be `list[Result[Any, Exception]]`? Same for `AllSettled`

File: src/resonate/actions.py (hunk 1 above)
Comment (avillega) [reply src/resonate/actions.py:109]: Where should we put them? should them be private methods of the scheduler?

File: src/resonate/actions.py (hunk 2 above)
Comment (avillega) [reply src/resonate/actions.py:137]: part of it is an artifact of trying to comply with the Combinators interface. the All combinator is right though, given that if one of the promises fail we want to return a single error. For all_settled I agree we could make it `list[Result[Any, Exception]]` once we make them private methods of the Scheduler

File: src/resonate/actions.py (hunk 1 above)
Comment (Tomperez98) [reply src/resonate/actions.py:109]: It believe so. This methods belong to the kernel space, and shouldn't face the user.

**Pull request merged**